File: algo/39. Combination Sum.js
// @9 DFS

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    var helper = function(res, collection, index, remain_target) {
        if (index > candidates.length) {
            return;
        }
        
        var collection = Array.from(collection);
        if (remain_target == 0) {
            res.push(collection);
        }
        
        for (var i = index; i < candidates.length; i++) {
            if (candidates[i] > remain_target) {
                // note that it should be continue here instead of break
                continue;
            }
            
            collection.push(candidates[i]);
            // since it allows using one number multiple times,
            // we parse i instead of i+1 into the recursive function call
            helper(res, collection, i, remain_target - candidates[i]);
            collection.pop();
        }
    }
    
    var res = [];
    if (candidates.length == 0) {
        return res;
    }
    
    candidates.sort();
    helper(res, [], 0, target);
    return res;
    
};

File: round2/39. Combination Sum.cs
public class Solution {
    private void backtrack(List<IList<int>> res, List<int> tempList, int[] nums, int target, int index) {
        if (target < 0) return;
        var tempList_cp = new List<int>(tempList);
        if (target == 0){
            res.Add(tempList_cp);
        } else {
            for (int i=index; i<nums.Length; i++) {
                tempList_cp.Add(nums[i]);
                backtrack(res, tempList_cp, nums, target-nums[i], i);
                tempList_cp.RemoveAt(tempList_cp.Count-1);
            }
        }
    }
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {       
        var res = new List<IList<int>>();
        if (candidates.Length==0) return res;
        Array.Sort(candidates);
        backtrack(res, new List<int>(), candidates, target, 0);
        return res;
    }
}